Command line parser plugins and ReadFilter command line arguments.

cmnbroad opened the pull request.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java:
```diff
@@ -71,6 +73,14 @@
 
+    // Map from (full class) name of each GATKCommandLinePluginDescriptor requested and
+    // found to the actual descriptor instance
+    Map<String, GATKCommandLinePluginDescriptor<?>> pluginDescriptors = new HashMap<>();
```
Can this be private rather than package-protected (since it has a public accessor)?

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java:
```diff
@@ -113,7 +123,7 @@ private boolean inArgumentMap(ArgumentDefinition arg){
 
     // List of all the data members with @Argument annotation
-    private final List<ArgumentDefinition> argumentDefinitions = new ArrayList<>();
+    List<ArgumentDefinition> argumentDefinitions = new ArrayList<>();
```
Here too -- why does this need to be package visible?

cmnbroad replied:
Done (private).

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java:
```diff
@@ -138,16 +148,26 @@ private String getUsagePreamble() {
+        this(callerArguments, null);
+    }
+
+    public CommandLineParser(
```
Add javadoc for public constructors

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/ReadWalker.java:
```diff
@@ -98,15 +105,50 @@ private SimpleInterval getReadInterval(final GATKRead read) {
+        // This filter gets returned if all others are disabled.
+        CountingReadFilter defaultFilter = new CountingReadFilter(
+                ReadFilterLibrary.ALLOW_ALL_READS.getClass().getSimpleName(),
+                ReadFilterLibrary.ALLOW_ALL_READS);
```
Can the `CountingReadFilter` get the class name internally, and simplify this to a 1-arg constructor?

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKCommandLinePluginDescriptor.java:
```diff
@@ -0,0 +1,142 @@
+    /**
+     * Package names from which to load the command line plugin classes
+      */
+    public abstract String getPackageName();
```
Can we make this return a `List<String>`, to allow future support for multiple packages per plugin type?

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/argumentcollections/ReadInputArgumentCollection.java:
```diff
@@ -26,6 +26,9 @@
     public ValidationStringency readValidationStringency = ReadConstants.DEFAULT_READ_VALIDATION_STRINGENCY;
 
+    @Argument(fullName = "disableAllReadFilters", shortName = "f", doc = "Disable all read filters", common = false, optional = true)
+    public boolean disableAllReadFilters = false;
```
Shouldn't this argument live in the descriptor, along with the other arguments that control selection/exclusion of read filters?

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -0,0 +1,199 @@
+    @Argument(fullName = disableReadFilterArgName,
+            shortName = "df",
+            doc="Read filters to be disabled before analysis", optional=true)
+    public final Set<String> disableFilters = new HashSet<>();
```
What if the user disables a filter that's on by default, but supplies argument(s) for the disabled filter?

cmnbroad replied:
This case will be detected/rejected now during final post-parse descriptor validation.

cmnbroad replied:
Rejected at descriptor validation time.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -0,0 +1,199 @@
+    @Argument(fullName = StandardArgumentDefinitions.READ_FILTER_LONG_NAME,
+            shortName = StandardArgumentDefinitions.READ_FILTER_SHORT_NAME,
+            doc="Read filters to be applied before analysis", optional=true)
+    public final List<String> readFilterNames = new ArrayList<>(); // preserve order
```
What if a tool has a filter on by default that takes arguments?

cmnbroad replied:
Done, with tests added.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -0,0 +1,199 @@
+
+    // Verify that this target plugin class's dependent value has been specified
+    @Override
+    public boolean isDependentArgumentAllowed(Class<?> targetPluginClass) {
```
I think this needs to become a way of registering plugin classes for which arguments were encountered. Then, after parsing is finished and the plugin system creates a final list of plugins selected (using both the selection/exclusion args and the tool defaults), you can do a second validation pass here to throw if args were specified for a plugin that wasn't selected.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKCommandLinePluginDescriptor.java:
```diff
@@ -0,0 +1,142 @@
+     * @param longArgName
+     * @return Set<String> of allowable values</String>, or null if any value is allowed
+     */
+    public abstract Set<String> getAllowedStringValues(String longArgName);
```
Needs better name perhaps? Something like `getAllowedValuesForDescriptorArgument()`

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKCommandLinePluginDescriptor.java:
```diff
@@ -0,0 +1,142 @@
+     *
+     * @param consumer function that accepts a collection of plugin instances
+     */
+    public abstract void getInstances(Consumer<Collection<T>> consumer);
```
Should this getter just return `Collection<T>`, instead of taking a `Consumer` and returning nothing?

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -0,0 +1,199 @@
+    public Class<?> getPluginClass() {return org.broadinstitute.hellbender.engine.filters.ReadFilter.class;}
+
+    @Override
+    public String getPackageName() {return "org.broadinstitute.hellbender.engine.filters";};
```
Declare package/plugin class as constants at the top

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -0,0 +1,199 @@
+    private ReadFilter getReadFilterForName(final String filterName) {
+        ReadFilter rf = null;
+        for (Map.Entry<String, ReadFilter> entry : readFilters.entrySet()) {
+            if (entry.getKey().endsWith(filterName)) {
```
Can you do a map lookup here?

cmnbroad replied:
Yes; the whole method is eliminated.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -0,0 +1,199 @@
+                requestedReadFilters.put(s, trf);
+            }
+        });
+        readFilters = requestedReadFilters;
```
I think all of this validation should be moved to a second pass that takes into account the tool-requested filters, as proposed above.

cmnbroad replied:
Tool defaults are passed in to the descriptor now before parsing, so this method can validate all inconsistencies.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -0,0 +1,199 @@
+     * @return
+     */
+    public <T extends ReadFilter> T getMergedReadFilters(
+            final List<ReadFilter> defaultFilters,
```
Rename `defaultFilters` to `toolFilters` to make it clear where they come from

cmnbroad replied:
These are passed in to the constructor now.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/ReadWalker.java:
```diff
@@ -98,15 +105,50 @@ private SimpleInterval getReadInterval(final GATKRead read) {
+
+        // Unless all filters are disabled, merge the tool's default filters with the users's command line
+        // read filter requests, then initialize the resulting filters and wrap each one in a CountingReadFilter.
+        return readArguments.disableAllReadFilters == true ?
```
`disableAllReadFilters` should be handled internally in the descriptor validation

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/ReadWalker.java:
```diff
@@ -98,15 +105,50 @@ private SimpleInterval getReadInterval(final GATKRead read) {
+                        f -> new CountingReadFilter(f.getClass().getSimpleName(), f),
+                        (f1 , f2) -> f1.and(f2),
+                        defaultFilter
+                );
```
Have a simple method that tool authors can call to simplify the task of defining default read filters, that assumes that filters should be wrapped in `CountingReadFilters`, should be `anded` together, etc. Tool authors can always do it manually if they need exotic behavior, but most can just call into the easy method.

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -11,41 +12,136 @@
+    public static ReadLengthEqualsCigarLength READLENGTH_EQUALS_CIGARLENGTH = new ReadLengthEqualsCigarLength();
+    public static SeqIsStored SEQ_IS_STORED = new SeqIsStored();
+    public static ValidAlignmentStart VALID_ALIGNMENT_START = new ValidAlignmentStart();
+    public static ValidAlignmentEnd VALID_ALIGNMENT_END = new ValidAlignmentEnd();
```
These should be `static final`

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/filters/FragmentLengthReadFilter.java:
```diff
@@ -3,14 +3,19 @@
+            shortName = "maxFragment",
+            doc = "Keep only read pairs with fragment length at most equal to the given value",
+            optional = false)
+    public Integer maxFragmentLength; // no default value, since the command line parser will think its set
```
Can we allow args in plugins to have default values? This arg, for example, should definitely be optional, as it was originally, as we don't want to require the user to always specify `maxFragmentLength` when using this filter. Should be possible if you add a callback to your descriptor at the point in the arg parsing where it injects values into filter instances, to keep track of which filters had args specified (needed for post-parsing validation when you build the final list of selected filters)

cmnbroad replied:
Yes, all the individual filters have been updated to have defaults where appropriate.

droazen commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -0,0 +1,199 @@
+    // Verify that this target plugin class's dependent value has been specified
+    @Override
+    public boolean isDependentArgumentAllowed(Class<?> targetPluginClass) {
+        return readFilterNames.contains(targetPluginClass.getSimpleName());
```
Filters requested by the tool and filters excluded should be taken into account here (or validation has to be delayed somehow). May need to have the arg parsing system notify via callback whenever it encounters an argument for a particular plugin class, so that you can later determine whether the plugin args the user specified were appropriate given the final selection of plugins.

cmnbroad replied:
Done, with tests added for these cases.

droazen commented:
Partial review complete -- will do a second pass after initial revisions.

droazen commented:
Can you make sure that all read filter args have long-ish/unique short and long names? Single-letter names should be reserved only for important universal GATK args, not plugin args.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/InsertSizeMetricsCollectorSpark.java:
```diff
@@ -119,18 +119,35 @@ public InsertSizeMetricsReadFilter(final InsertSizeMetricsArgumentCollection.End
+            if(0!=MQThreshold)  { tempFilter = tempFilter.and(new ReadFilter() {
+                private static final long serialVersionUID = 1L;
+                @Override public boolean test(final GATKRead read){return read.getMappingQuality() >= MQThreshold;}});};
 
```
Many of these inline filters can be replaced with ReadFilterLibrary instances.

cmnbroad commented:
Rather than delaying argument validation to a second pass, I changed the sequence so the tool instantiates the descriptor and gives it the tool defaults right from the start, (and then passes the descriptor instance(s) to the arg parser) so it has all the state it needs to validate at arg parsing time. Also, I reverted the removal of the package limitation for plugins, at least temporarily, since searching through all packages looked like it slowed down the integration tests.

cmnbroad commented:
Let me know if you'd prefer a squash/rebase before review.

droazen commented:
@lbergelson Would you be able to do the second-pass review here, since I did the first pass? In particular, it would be good if you could scrutinize the changes to the arg parser.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java:
```diff
@@ -138,16 +148,38 @@ private String getUsagePreamble() {
 
-        createArgumentDefinitions(callerArguments);
+        createArgumentDefinitions(callerArguments, null);
+        if (pluginDescriptors != null) {
```
I think we should generally prefer passing an empty list vs a null for something like this.

cmnbroad replied:
Done.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java:
```diff
@@ -138,16 +148,38 @@ private String getUsagePreamble() {
+    public CommandLineParser(
+            final Object callerArguments,
+            final List<? extends GATKCommandLinePluginDescriptor<?>> pluginDescriptors) {
         this.callerArguments = callerArguments;
```
While you're here could you add null checks for callerArguments?

cmnbroad replied:
Done.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java:
```diff
@@ -138,16 +148,38 @@ private String getUsagePreamble() {
     }
 
-    private List<ArgumentDefinition> createArgumentDefinitions(final Object callerArguments) {
+    private void createArgumentDefinitions(
```
good catch,  don't know how this ended up with a return type...

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java:
```diff
@@ -157,20 +189,56 @@ public CommandLineParser(final Object callerArguments) {
+    // Find all the instances of plugins specified by the provided plugin descriptors
+    private void createCommandLinePluginArgumentDefinitions(
+            final List<? extends GATKCommandLinePluginDescriptor<?>> requestedPluginDescriptors) {
+        // For each descriptor, create an argument definition for the descriptor itself, then
```
should this read "create argument definitions" instead of being singular?

cmnbroad replied:
At first I thought not, but I see your point; there can be multiple defs for the descriptor itself. Updated.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java:
```diff
@@ -195,21 +263,41 @@ public void usage(final PrintStream stream, final boolean printCommon) {
+            // group the argdefs by their parent (plugin) class
+            Map<String, List<ArgumentDefinition>> groupedArgsMap = conditionalArgs.stream()
+                    .collect(Collectors.groupingBy(argDef -> argDef.parent.getClass().getSimpleName()));
+            stream.println("\nConditional Arguments:\n");
```
This is nitpicky, but it would be nice if instead of just "Conditional Arguments" these were grouped by plugin descriptor.  I.e. All the read filters are together in a block that says "Read Filter Arguments", and a second plugin would be listed separately.  We could patch that in a later pull request if you want though.

lbergelson replied:
Also, could we sort these alphabetically before printing them?

cmnbroad replied:
Grouped and sorted, both. While I was at it I sorted the required and optional args as well.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java:
```diff
@@ -349,7 +437,50 @@ private void assertArgumentsAreValid() {
+            if (null == argumentDefinition.controllingDescriptor) {
+                actualArgumentDefinitions.add(argumentDefinition);
+            }
+            else {
```
there's funny line breaking in this block and the one following it, we usually go with` } else {` I think.

cmnbroad replied:
Done.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java:
```diff
@@ -349,7 +437,50 @@ private void assertArgumentsAreValid() {
+    private void validatePluginArguments() {
+        final List<ArgumentDefinition> actualArgumentDefinitions = new ArrayList<>();
+        for (final ArgumentDefinition argumentDefinition : argumentDefinitions) {
+            if (null == argumentDefinition.controllingDescriptor) {
```
Could you extract a method that does this check and name it to make it more clear what this is testing for, maybe `isPluginArgument`.

cmnbroad replied:
Done.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java:
```diff
@@ -493,7 +624,13 @@ private String makeArgumentDescription(final ArgumentDefinition argumentDefiniti
             sb.append("Required. ");
         }
-        sb.append(getOptions(getUnderlyingType(argumentDefinition.field)));
+        if (GATKCommandLinePluginDescriptor.class.isAssignableFrom(argumentDefinition.parent.getClass()) &&
```
Please add a comment here explaining this check.

cmnbroad replied:
Done.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java:
```diff
@@ -513,10 +649,22 @@ private String makeArgumentDescription(final ArgumentDefinition argumentDefiniti
+        final GATKCommandLinePluginDescriptor<?> descriptor = (GATKCommandLinePluginDescriptor<?>) argDef.parent;
+        // this argument came from a plugin descriptor; delegate to get the list of allowed values
+        final Set<String> allowedValues = descriptor.getAllowedValuesForDescriptorArgument(argDef.getLongName());
+        if (allowedValues == null) {
```
Usually we'd want to prefer empty collections over null values, is there a need to distinguish the two here

cmnbroad replied:
Nope. Changed the interface javadoc and the code to used an empty list.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java:
```diff
@@ -513,10 +649,22 @@ private String makeArgumentDescription(final ArgumentDefinition argumentDefiniti
+        if (allowedValues == null) {
+            sb.append("Any value allowed");
+        } else {
+            sb.append("Allowed values are: ");
```
Sort these so they're output in alphabetical order (or if you want the plugin descriptor to control the order sort them there.

cmnbroad replied:
Done.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKCommandLinePluginDescriptor.java:
```diff
@@ -0,0 +1,159 @@
+ *  getAllowedValuesForDescriptorArgument is only called when the command line parser is constructing
+ *  a help/usage message.
+ */
+public abstract class GATKCommandLinePluginDescriptor<T> {
```
We might want to add a `getPluginName()` to support printing the plugin args in an even more user friendly way.

cmnbroad replied:
Yeah. Done.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -0,0 +1,334 @@
+            doc="Read filters to be applied before analysis", optional=true)
+    public final List<String> userReadFilterNames = new ArrayList<>(); // preserve order
+
+    final String disableReadFilterArgName = "disableReadFilter";
```
This should probably be static

cmnbroad replied:
Done.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -0,0 +1,334 @@
+
+    final String disableReadFilterArgName = "disableReadFilter";
+    @Argument(fullName = disableReadFilterArgName,
+            shortName = "df",
```
It's strange that this is lower case but read filter is 'RF'

cmnbroad replied:
Good point. Made it upper case.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -0,0 +1,334 @@
+        // throw if any filter is both enabled *and* disabled by the user
+        final Set<String> enabledAndDisabled = new HashSet<>(userReadFilterNames);
+        enabledAndDisabled.retainAll(disableFilters);
+        enabledAndDisabled.forEach(
```
It's a bit strange to have this as a forEach since it can really only happen once.

cmnbroad replied:
Agreed. fixed.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -0,0 +1,334 @@
+    /**
+     * Determine if a particular ReadFilter was disabled on the command line.
+     * @param filterName name of the filter to query
+     * @return true if the name appears the list of disabled filters
```
typo: true if the name appears **in** the list of disabled filters

cmnbroad replied:
Done.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/LocusWalker.java:
```diff
@@ -65,20 +64,52 @@ public boolean includeNs() {
+        // read filter requests, then initialize the resulting filters and wrap each one in a CountingReadFilter.
+        GATKReadFilterPluginDescriptor readFilterPlugin =
+                (GATKReadFilterPluginDescriptor)
+                        commandLineParser.getPluginDescriptor(GATKReadFilterPluginDescriptor.class);
```
I think it would be nicer if the cast was pushed into `getPluginDescriptor`

cmnbroad replied:
Done.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java:
```diff
@@ -71,6 +73,14 @@
+    private Map<String, GATKCommandLinePluginDescriptor<?>> pluginDescriptors = new HashMap<>();
+
+    // Return the plugin instance corresponding to the targetDescriptor class
+    public GATKCommandLinePluginDescriptor<?> getPluginDescriptor(Class<?> targetDescriptor) {
```
I commented below, about this, but I think it would probably be better add a type parameter to this and have it cast the result internally to the target descriptor type.

cmnbroad replied:
Good idea. Done.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java:
```diff
@@ -11,41 +12,136 @@
 
     //Note: do not call getCigar to avoid creation of new Cigar objects
-    public static final ReadFilter READLENGTH_EQUALS_CIGARLENGTH = read -> read.isUnmapped() || read.getLength() == Cigar.getReadLength(read.getCigarElements());
+    public static class CigarIsSupported extends ReadFilter {
```
Should these all have ReadFilter as part of their name?  It's a bit inconsistent on the commandline that some do and some don't.

cmnbroad replied:
Yes. Done.

lbergelson commented:
What happens if an inactive read filter has the same argument as one of the arguments in a tool, is there an error, or only if there is an active read filter with a name collision? 
 What if two active read filters have the same argument name?  Are there tests for name collision?  I didn't see them but I might have missed it.

lbergelson commented:
@cmnbroad Back to you.  Some minor cosmetic comments about printing the help output.  I think it looks good though.  This version with the plugin descriptors is so much nicer than the early version.

cmnbroad commented:
Back to @lbergelson - not sure if you wanted to re-review.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java:
```diff
@@ -1023,6 +1054,13 @@ public boolean isFlag(){
         }
 
+        /**
+         * Determine if this argument definition is controlled by a plugin descripto (and thus subject to
```
missing r in descripto

cmnbroad replied:
Done.

lbergelson commented:
@cmnbroad Looks good!  Lets get this merged finally.  Rebase and merge at will.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -0,0 +1,337 @@
+
+    final static String disableReadFilterArgName = "disableReadFilter";
+    @Argument(fullName = disableReadFilterArgName,
+            shortName = "DF",
```
Sorry for the meddling, but while trying to understand how a plugin will be implemented I found this: the names for the read filter are in `StandardArgumentDefinitions`, the disable read filter in a static field (only the long name, the other is hardcoded in the annotation), and the one for disable all the read filters is inside the annotation. I think that it would be nice to have all this values in `StandardArgumentDefinitions` to be able to have a class to check possible collisions...

cmnbroad replied:
Fair enough - its an artifact left over from the first implementation. Done.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -0,0 +1,337 @@
+    public String getDisplayName() { return StandardArgumentDefinitions.READ_FILTER_LONG_NAME; }
+
+    /**
+     * @return the class object for the base class of all plugins managed by thid descriptor
```
Again, sorry for being picky: `thid` -> `this`

cmnbroad replied:
Done

cmnbroad commented:
Also, good point on the arg name collision. The command line parser handles this (for any args, even across plugins or across anywhere) but I didn't see any tests, so I added tests to both CommandLineParserUnitTest and CommandLineParserPluginUnitTest.
